[KAN-7] : Created responsive-header-component

Uzmakh opened the pull request:
Hi Peers, 
       I have created a mobile sidebar nav-menu as well as the full-width desktop responsive navbar within a header. Though I 
 still have some steps incomplete in my code, they can be refurbished soon. Here are the screenshots of the outcome;
<img width="957" alt="myteam-res-header-1" src="https://github.com/poulami-saha/Global-Code-Queens-MyTeam-Website/assets/91914613/a70d4804-114a-4d5d-90b6-3eaec8169fb0">
<img width="595" alt="myteam-res-header-2" src="https://github.com/poulami-saha/Global-Code-Queens-MyTeam-Website/assets/91914613/67c3d228-516e-43a2-aaf1-55c259db4d40">
<img width="545" alt="myteam-res-header-3" src="https://github.com/poulami-saha/Global-Code-Queens-MyTeam-Website/assets/91914613/6f7aa44d-74d0-47b2-b9dd-443487ae9868">
<img width="375" alt="myteam-res-header-4" src="https://github.com/poulami-saha/Global-Code-Queens-MyTeam-Website/assets/91914613/8474412c-0d98-4e2a-8499-c95c23d7238c">

Uzmakh committed:
Created responsive-header-component

Uzmakh committed:
Added backgroung image to mobile-view side menu-bar

poulami-saha commented on src/components/Header/Header.jsx:
```diff
@@ -1,17 +1,27 @@
 
 const Header = () => {
+  const [mobileMenu, setMobileMenu] = useState(false);
+  const toggleMenu = () => {
```
```suggestion
 const toggleMenu = () => {
    setMobileMenu((toggleMenu) => !toggleMenu);
  };
```

poulami-saha replied:
We just need to reverse the value so this can be of less code

joanwolk replied:
See https://github.com/poulami-saha/Global-Code-Queens-MyTeam-Website/pull/24/commits/49b796c9c657122a39dbfcac09176ec80a692af9#r1685844601

poulami-saha commented on src/components/Header/Header.jsx:
```diff
@@ -1,17 +1,27 @@
-          <li><Link to="/contact">Contact Us</Link></li>
+        <ul className={ mobileMenu ? '' :  styles['hide-mobile-menu']}>
+          <img src={icon_close} alt="" className={styles.icon_close} onClick={toggleMenu} />
+         <li>home</li>
```
```suggestion
 const navBarOptions = ["home", "about"];
        {navBarOptions.map((option) => {
            return (
              <li key={option}>
                <Typography elType="body3">{option}</Typography>
              </li>
            );
          })}
```

poulami-saha replied:
We can add eltype body3 with the following details
`.body3 {
  font-size: 1.12rem;
  line-height: 1.75rem;
}`

poulami-saha replied:
now its 2, may be some more in future, we make an array of route name and rendered component and iterate over it in the JSX code

joanwolk replied:
Please re-add the `Link` components to the nav bar—right now it doesn't navigate.
`<Link to="/">home</Link>`
`<Link to="/about">About</Link>`

joanwolk commented on src/components/Header/Header.jsx:
```diff
@@ -1,17 +1,27 @@
-import { Link } from 'react-router-dom';
-import styles from './Header.module.css';  
+import React, { useState } from "react";
+// import { Link } from 'react-router-dom';
```
Please re-enable the Link import

joanwolk commented on src/components/Header/Header.jsx:
```diff
@@ -1,17 +1,27 @@
+         <li>home</li>
+          <li>about</li>
         </ul>
+          <button className={styles.btn}>contact us</button>
```
Please wrap the button in a `<Link to="/contact">` to enable navigation. This will get it working as navigation, and that's enough for us on this project right now.

Ideally, for semantic HTML and best-practice accessibility, we would style the link without an actual `button`, but for us for now, we'll gloss over that to get the component working and looking right. If we have time later, we can look at our best option for this.

berikawilliams commented:
I noticed that the header component has a css style for the .header class as “position:fixed”, which is causing content on the home and about page to be partially hidden at the top. Was your intent to create a fixed header when the user scrolls down?

Uzmakh commented:
@berikawilliams yess, this is a fixed header when user scrolls down. If content is to be hidden, we fix it. Thankyou

Uzmakh committed:
Reverted Links and set the side menubar

Uzmakh committed:
Kept the button-contact-us outside of <ul> but inside <nav>

joanwolk commented on src/components/Header/Header.module.css:
```diff
@@ -1,20 +1,128 @@
+.logo {
+  width: 160px;
+}
+nav {
```
Please use classes to style with, not the element types. CSS modules will automatically adjust **class names** to prevent styles meant for one place getting accidentally applied on a matching selector outside the component (where it doesn't belong), but it _can't_ do that for element names.

Suggested class name for `nav` element: `.main-nav` or `.header-nav`

joanwolk commented on src/components/Header/Header.jsx:
```diff
@@ -1,18 +1,32 @@
+         
+          {/* <button className={styles.btn}>contact us</button> */}
         </ul>
+         <li><Link to="/contact">Contact Us</Link></li>
```
An `<li>` tag (list item) should always be inside either an `<ol>` (ordered list, aka numbered list) or a `<ul>` (unordered list, aka bullet point list).

To style the "Contact Us" link, you can use CSS to have the final `li` in the list align right instead of left.

joanwolk commented on src/components/Header/Header.module.css:
```diff
@@ -1,20 +1,128 @@
+  align-items: center;
+  justify-content: space-between;
+}
+nav ul {
```
suggested class name for the `nav ul`: `.nav-list`

joanwolk commented on src/components/Header/Header.module.css:
```diff
@@ -1,20 +1,128 @@
+  align-items: center;
+}
+
+nav ul li {
```
suggested class name: `.nav-list-item`

Assuming the `ul` takes up the full available space (hint: use width 100% because it's flex), the way to get the last `li` to align right would be something like
```
.nav-list-item:last-child {
  margin-left: auto;
}
```

joanwolk commented on src/components/Header/Header.module.css:
```diff
@@ -1,20 +1,128 @@
+
+nav ul li {
+  list-style: none;
+  color: var(--primary-white);
```
Because you're styling links (`<a>`, anchor tags), you'll need to apply color to the `a` rather than inheriting it from the parent.

With CSS modules, I would probably add a class on the `Link` component called something like `.nav-link` and add the color rules on that. (Don't forget a separate selector for `.nav-link:hover` to set the hover color!)

```
.nav-link {
  color: var(--primary-white);
  text-decoration: none; /* gets rid of the underline */
}

.nav-link:hover {
  color: var(--coral-color);
}
```

Uzmakh committed:
Used classes to style with

joanwolk commented on src/components/Header/Header.jsx:
```diff
@@ -1,18 +1,32 @@
-          <li><Link to="/about">About</Link></li>
-          <li><Link to="/contact">Contact Us</Link></li>
+      <img src={logo} alt="myteam logo" className={styles.logo} />
+      <nav className={styles.header-nav}>
```
CSS Modules don't support calling hyphenated (aka "kebab-case") class names like `styles.header-nav`. If you use hyphenated class names, use the form `styles['header-nav']` to call the styles. 

OR you can use camelCase class names instead: `styles.headerNav` will work great. This tends to be how React devs do it, to the best of my knowledge.

joanwolk replied:
Apologies for suggesting kebab-case-names when that's not the most common in React. I saw that you already had one in the file, so I just went with that rather than trying to rename that immediately, and I would do it differently next time.

joanwolk commented on src/components/Header/Header.jsx:
```diff
@@ -1,18 +1,32 @@
+        <ul className={ mobileMenu ? '' :  styles['hide-mobile-menu']}>
+          <img src={icon_close} alt="" className={styles.icon_close} onClick={toggleMenu} />
+       <li className={styles.nav-list-item}><Link to="/">Home</Link></li>
+          <li className={styles.nav-list-item}><Link to="/about">About</Link></li>
```
same issue as above with hyphenated class name syntax

please also fix lines 20 and 22 to align correctly with the indentation level of line 21

joanwolk commented on src/components/Header/Header.module.css:
```diff
@@ -16,26 +16,35 @@
+.nav-link {
+  color: var(--primary-white);
+  text-decoration: none;
+  /* gets rid of the underline */
```
I would either put this comment at the end of line 41 (clearly refers to the `text-decoration` rule) or else remove the comment. I had it in the sample code to be obvious about what it was doing, in case that was unfamiliar, but once it's on its own line, the odds of it getting separated from the relevant line or becoming unclear about what it refers to go way up.

joanwolk commented on src/components/Header/Header.jsx:
```diff
@@ -14,15 +14,15 @@ const Header = () => {
-         
+       <li className={styles.nav-list-item}><Link to="/">Home</Link></li>
+          <li className={styles.nav-list-item}><Link to="/about">About</Link></li>
+           <li className={styles.nav-list-item}><Link to="/contact">Contact Us</Link></li>
```
apply the class to the `Link` component to get the colors where you want them:
`<Link to="/" className={styles['nav-link']}>`
`<Link to="/about" className={styles['nav-link']}>`
`<Link to="/contact" className={classNames(styles['nav-link'], styles.btn)}>` (see [other comment](https://github.com/poulami-saha/Global-Code-Queens-MyTeam-Website/pull/24/files#r1681830893) about classNames library)

(or with `styles.navLink` if you rename to use camelCase)

joanwolk commented on src/components/Header/Header.jsx:
```diff
@@ -1,18 +1,32 @@
-          <li><Link to="/contact">Contact Us</Link></li>
+      <img src={logo} alt="myteam logo" className={styles.logo} />
+      <nav className={styles.header-nav}>
+        <ul className={ mobileMenu ? '' :  styles['hide-mobile-menu']}>
```
This `ul` needs to be `display: flex;` and some other stuff, so we need to have the styles you already added as `.nav-list` get applied here _all the time_. So we need to conditionally add multiple classes. The classNames library is already doing a similar thing in the Button utility component; you can take a look [here](https://github.com/poulami-saha/Global-Code-Queens-MyTeam-Website/blob/6f78123f16138af0b45ce0bc4fffb3e7a1270b7a/src/components/utility/Button/Button.jsx#L22-L24).

Add `import classNames from "classnames";` in the imports at the top of the file

then after the mobile menu toggle and before the JSX, add this:
```
  const listClass = classNames(styles['nav-list'], {
    [styles['hide-mobile-menu']]: mobileMenu,
  });
```

joanwolk commented on src/components/Header/Header.module.css:
```diff
@@ -1,20 +1,137 @@
+.nav-list-item{
+  list-style: none;
+  color: var(--primary-white);
+  padding-right: 30px;
```
Remove the `padding-right`. You were looking for `margin-right` (space outside the element) but since we're using flex, we can use the `gap` property on the parent (`.nav-list` in this case) to easily set space _only_ between elements and not add extra spacing on the end (the right end, in this case).

joanwolk commented on src/components/Header/Header.module.css:
```diff
@@ -1,20 +1,137 @@
+}
+.nav-list {
+  display: flex;
+  align-items: center;
```
Add `gap: 30px;` to put spacing between `.nav-list-item`s

joanwolk replied:
Add `width: 100%;` to prepare for the neatest way to position the contact link

joanwolk commented on src/components/Header/Header.module.css:
```diff
@@ -1,20 +1,137 @@
+  list-style: none;
+  color: var(--primary-white);
+  padding-right: 30px;
+}
```
To get the last list item (the contact link) to align on the right edge, add 
```
.nav-list-item:last-child {
  margin-left: auto;
}
```
and as long as the ul fills the width here, it will line up perfectly

Uzmakh committed:
Updated suggested changes

joanwolk committed:
fix class names to work with CSS Modules and classNames library

joanwolk committed:
Merge branch 'main' into responsive-header-component

joanwolk committed:
fix toggle function of mobile navigation
- pretty sure this could get refactored to re-combine the nav and mobile nav menus, but this at least works now

joanwolk commented on src/components/Header/Header.jsx:
```diff
@@ -1,5 +1,6 @@
 import React, { useState } from "react";
-import { Link } from "react-router-dom";
+import { Link } from 'react-router-dom';
+import classNames from "classnames";
```
This import is required to make lines 14 and 31 work correctly. Confusingly, `classNames` is a library, while `className` (singular) is a property built into React/JSX

joanwolk commented on src/components/Header/Header.jsx:
```diff
@@ -16,29 +17,18 @@ const Header = () => {
+      <nav className={styles.headerNav}>
+        <ul className={listClass}>
+          <img src={icon_close} alt="" className={styles.icon_close} onClick={toggleMenu} />
+          <li className={styles['nav-list-item']}>
```
Hyphenated class names like `nav-list-item` need to use this format, which _MUST_ include quotation marks (e.g., `styles["nav-list-item"]`; either single or double will work) for at least two reasons:
1. Without quotation marks, it won't correctly match the thing it's looking for
2. Without quotation marks, automatic formatting tools can break them when trying to make math operations readable, as you can see in the diff: the formatter turned `styles[nav-list-item]` into `styles[nav - list - item]`, which is broken because the key is pulled apart into multiple pieces

joanwolk commented on src/components/Header/Header.jsx:
```diff
@@ -2,36 +2,68 @@ import React, { useState } from "react";
-        </ul>
-      </nav>
+      {standardNavMenu}
+      {mobileNavMenu}
```
I separated standard nav and mobile nav while working on the refactor to get display of mobile nav working correctly. There is probably a way to re-combine these in a further refactoring, which I will also work on. Having them separate may still be helpful while refactoring to have the mobile styles first and the tablet/desktop media queries after.

joanwolk commented on src/components/Header/Header.jsx:
```diff
@@ -2,36 +2,68 @@ import React, { useState } from "react";
+  })
+
+  const navLinks = (
+    <>
```
This empty-looking `<>` tag opens a React Fragment, and is useful for returning one containing element to React without adding extra `div` tags to the HTML. The closing tag is `</>` and can be seen on line 39.

joanwolk commented on src/components/Header/Header.jsx:
```diff
@@ -2,36 +2,68 @@ import React, { useState } from "react";
+  const [mobileMenuOpen, setMobileMenuOpen] = useState(false);
   const toggleMenu = () => {
-    setMobileMenu((toggleMenu) => !toggleMenu);
+    setMobileMenuOpen(!mobileMenuOpen);
```
I ended up changing the name of the state and its setter, but the important thing I changed here is that **it's just providing the new value** (the opposite boolean value), not running a function (and especially not running a function that returns `void`)

joanwolk committed:
revise header color and padding

joanwolk commented:
Merged #35 instead
